dev/devicelab/lib/framework/host_agent.dart

```diff
@@ -0,0 +1,122 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+
+import 'package:file/file.dart';
+import 'package:file/local.dart';
+import 'package:meta/meta.dart';
+import 'package:platform/platform.dart';
+
+/// The current host machine running the tests.
+HostAgent get hostAgent => HostAgent(platform: const LocalPlatform(), fileSystem: const LocalFileSystem());
+
+/// Host machine running the tests.
+abstract class HostAgent {
+  factory HostAgent({@required Platform platform, @required FileSystem fileSystem}) {
+    if (platform.isWindows) {
+      return _WindowsHostAgent(platform: platform, fileSystem: fileSystem);
+    } else if (platform.isMacOS) {
+      return _MacOSHostAgent(platform: platform, fileSystem: fileSystem);
+    } else {
+      return _PosixHostAgent(platform: platform, fileSystem: fileSystem);
+    }
+  }
+
+  HostAgent._({@required Platform platform, @required FileSystem fileSystem})
+      : _platform = platform,
+        _fileSystem = fileSystem;
+
+  final Platform _platform;
+  final FileSystem _fileSystem;
+
+  /// Creates a directory to dump file artifacts.
+  Directory get dumpDirectory {
+    if (_dumpDirectory == null) {
+      // Set in LUCI recipe.
+      final String directoryPath = _platform.environment['FLUTTER_LOGS_DIR'];
+      if (directoryPath == null) {
+        _dumpDirectory = _fileSystem.systemTempDirectory.createTempSync('flutter_test_logs.');
+        print('Created tmp dump directory ${_dumpDirectory.path}');
+      } else {
+        _dumpDirectory = _fileSystem.directory(directoryPath)..createSync(recursive: true);
+        print('Found FLUTTER_LOGS_DIR dump directory ${_dumpDirectory.path}');
+      }
+    }
+    return _dumpDirectory;
+  }
+
+  static Directory _dumpDirectory;
+
+  @visibleForTesting
+  void resetDumpDirectory() {
+    _dumpDirectory = null;
+  }
+
+  Future<bool> dump();
+}
+
+class _WindowsHostAgent extends HostAgent {
+  _WindowsHostAgent({@required Platform platform, @required FileSystem fileSystem})
+      : super._(platform: platform, fileSystem: fileSystem);
+
+  @override
+  Future<bool> dump() async => true;
+}
+
+class _MacOSHostAgent extends HostAgent {
+  _MacOSHostAgent({@required Platform platform, @required FileSystem fileSystem})
+      : super._(platform: platform, fileSystem: fileSystem);
+
+  @override
+  Future<bool> dump() async {
+    // Copy simulator logs and crash reports.
+    final String home = _platform.environment['HOME'];
+    if (home == null) {
+      print(r'$HOME not found, skipping simulator log dump.');
+      return false;
+    }
+
+    final Directory simulatorLogs = _fileSystem.directory(_fileSystem.path.join(home, 'Library', 'Logs', 'CoreSimulator'));
```

Comment:
Is there a risk of us dumping old logs on a particular test with this sort of host based structure?

i.e. an agent runs a simulator task, then crashes running a physical ios test. The dumped logs include both combined.

Reply:
or I guess this could happen from the same agent - do these logs get cleaned up?

Reply:
[I asked this same question.](https://flutter-review.googlesource.com/c/recipes/+/9941/1/recipes/devicelab/devicelab_drone.py#42) 😄 

Me:
> Does this directory need to be removed after upload so the logs don't bleed between tests?

@godofredoc:
> The good news is that directory is the in the api.path['cleanup'] which luci cleans up automatically after the task execution.

As for these particular simulator logs, they actually aren't cleaned up between tests, so we'd need to correlate timestamps.  For example, the [example task execution](https://luci-milo.appspot.com/raw/build/logs.chromium.org/flutter/led/godofredoc_google.com/45dfad004f2463490eb571c2d74cba40e910aca07200f16f6aca2f3f7c258553/+/annotations) at  started at Jan 22 18:26 but shows simulator logs from Jan 17 18:08:47 (I've since removed the `system.log` files from upload, but it does upload app crashes from the simulator on task failure).  I can file a recipe issue to remove this directory before the Xcode select step, but I don't think this test runner code should worry about that.

However if we say uploaded a `adb logcat` with a timestamp of the beginning of the test and saved it to the directory, it would be cleaned up after task execution and wouldn't be present in any other tasks.

Reply:
something, something great minds.

I'd be concerned about leaving all of the responsibility of deleting old logs (outside of the upload dir) to the recipes. If someone less familiar with our CI system (most of us) wants to add new logs, then they would also need to know the change the recipe in the right way.

Checking timestamps here seems like a reasonable approach. And yeah, if logs come through adb with a timestamp start they should be fairly up to date.

Reply:
> If someone less familiar with our CI system (most of us) wants to add new logs, then they would also need to know the change the recipe in the right way.

I hope that would be rare and these artifacts would be more transient, like screenshots, or device log dumps, or IDE test logs, or size analysis files.  We could push more of that stuff into this upload directory instead of dumping it all into the task stdout as we do now.  I just happened to think of simulator app crashes first, and I had hoped they would be cleaned up between Xcode installs, but I guess not.

Reply:
what I would like is for this to fail in a way that doesn't upload logs/files instead of uploading old logs/files, because the former is much easier to notice immediately

Reply:
Are those logs generated in deriveddata? if yes, deriveddata is cleaned up before running every task.

Reply:
> Are those logs generated in deriveddata? if yes, deriveddata is cleaned up before running every task.

Unfortunately not.

Reply:
This has been removed.